libraries/AP_UAVCAN/AP_UAVCAN.cpp

```diff
@@ -1006,29 +1049,52 @@ AP_UAVCAN::Mag_Info *AP_UAVCAN::find_mag_node(uint8_t node)
 }
 
 /*
- * Find discovered not taken mag node with smallest node ID
+ * Find discovered mag node with smallest node ID and which is taken N times,
+ * where N is less than its maximum sensor id.
+ * This allows multiple AP_Compass_UAVCAN instanses listening multiple compasses
+ * that are on one node.
  */
 uint8_t AP_UAVCAN::find_smallest_free_mag_node()
 {
     uint8_t ret = UINT8_MAX;
 
     for (uint8_t i = 0; i < AP_UAVCAN_MAX_MAG_NODES; i++) {
-        if (_mag_node_taken[i] == 0) {
+        if (_mag_node_taken[i] < _mag_node_max_sensorid_count[i]) {
             ret = MIN(ret, _mag_nodes[i]);
         }
     }
 
     return ret;
 }
 
-void AP_UAVCAN::update_mag_state(uint8_t node)
+void AP_UAVCAN::update_mag_state(uint8_t node, uint8_t sensor_id)
 {
     // Go through all listeners of specified node and call their's update methods
     for (uint8_t i = 0; i < AP_UAVCAN_MAX_MAG_NODES; i++) {
         if (_mag_nodes[i] == node) {
             for (uint8_t j = 0; j < AP_UAVCAN_MAX_LISTENERS; j++) {
                 if (_mag_listener_to_node[j] == i) {
-                    _mag_listeners[j]->handle_mag_msg(_mag_node_state[i].mag_vector);
+                    
+                    /*If the current listener has default sensor_id,
+                      while our sensor_id is not default, we have
+                      to assign our sensor_id to this listener*/ 
+                    if ((_mag_listener_sensor_ids[j] == 0) && (sensor_id != 0)) {
+                        bool already_taken = false;
+                        for (uint8_t k = 0; k < AP_UAVCAN_MAX_LISTENERS; k++) 
```

Comment:
missing { } around for and if. I know it's just a single line so the logic is correct, but for clarity we like to always add them.